### DISCUSSION TITLE
Is one worker can only assigned to one player?

### DISCUSSION THREAD

**Discussion by gegaojian-richard** [giongto35/cloud-game#267] [Q&A] [answered]
Awesome work！In your design document and the code of worker, it seems one worker can handler more than one user, and each user session will spawn a new room running a gaming emulator.

But the following code make me confused, Is one worker can only assigned to one player?

```go
func (o *Server) WS(w http.ResponseWriter, r *http.Request) {
	log.Println("Coordinator: A user is connecting...")
	// ...

	// Assign available worker to browserClient
	bc.WorkerID = wc.WorkerID
	wc.IsAvailable = false

	// Everything is cool
	// Attach to Server instance with sessionID
	o.browserClients[sessionID] = bc
	defer o.cleanBrowser(bc, sessionID)

	// Routing browserClient message
	o.RouteBrowser(bc)

	bc.Send(cws.WSPacket{
		ID:   "init",
		Data: createInitPackage(wc.StunTurnServer, o.library.GetAll()),
	}, nil)

	// If peerconnection is done (client.Done is signalled), we close peerconnection
	<-bc.Done

	// Notify worker to clean session
	wc.Send(cws.WSPacket{
		ID:        "terminateSession",
		SessionID: sessionID,
	}, nil)

	// WorkerClient become available again
	wc.IsAvailable = true
}
```
A worker will not release until the user session quit.

**Comment by giongto35** [comment, marked as answer]
Hi @gegaojian-richard, 

This logic is checking if a new user joins an existing room, the worker handling that room will serve that user. If not, the user will pick among available server.

I think you got confused by this line "wc.IsAvailable = false" and "wc.IsAvailable = true" at the end of peerconnection. It looks like a bug here, wc.IsAvailable can only set to true if there is no peerconnection assign to it. 

Thanks very much for pointing that out!

**Reply by gegaojian-richard** [reply]
If i just have one coordinator and one worker, then i can only serve one player? (I have no available worker for another player because the only worker was assigned and that player hasn't quit)

But the design of worker is that it can create a new room for another player. This makes me confused, since coordinator will not assign this worker to another new player. 

Is it a bug, that wc.IsAvailable will not set to true until the end of peerconnection?

**Reply by giongto35** [reply]
"But the design of worker is that it can create a new room for another player"
Originally, one instance can run multiple workers. Each worker can have multiple rooms. Each room can host a single game and multiple users can connect to the same room running the same game.
However, after time, because of a limitation in libretro, each worker can run only one room, so you can assume worker-room is 1-1 mapping now. And if a user is assigned to a room, he is also assigned to a worker.

"Is it a bug, that wc.IsAvailable will not set to true until the end of peerconnection?": Yes, it's a bug. 

The code for this part looks confusing indeed. Thanks for raising those points. Please let me know if you have any further questions :

**Reply by gegaojian-richard** [reply]
I have made a try to remove the line "wc.IsAvailable = false" and "wc.IsAvailable = true" at the end of peerconnection.

It works well when i setup one coordinator and one worker, and the worker can really serve more than one player.  This conflicts with "each worker can run only one room".

It also works well when two players share one room in a worker, but it goes wrong when another player comes in(not in share mode). Is this the limitation in libretro?

**Reply by sergystepanov** [reply]
@gegaojian-richard, it's unlikely, that one worker would be able to emulate more than one game. Don't know about Libretro limitations, but internally, in a worker, there is just one shared Libretro emulator instance which uses (not isolated) global state. Even if you hack these flags, a worker wouldn't work properly with multiple games and it'd most likely just crash with a segfault.
So right now it's one worker to one game and that game can have up to 6 players, I think, plus any number of spectators.
If you want more games just run more workers, that's just 40 megs of RAM. Personally, I can't see any benefits in allowing workers to serve more games. If that worker crashes -- that will be bad not for one room but for many.

**Reply by giongto35** [reply]
When we mention worker, worker is just a single process. In one machine, we can run multiple worker processes. So we achieve one machine can run multiple game. And as Sergey said, it turns out to be a good design as well because we achieve full between each game session. 
Talkin about Libretro limitation. Before libretro, we can spawn multiple NES game sessions from a single process, so there can be many rooms in the same worker process. When I integrate with Libretro, Libretro game occupies global state and runs in the main thread, so I can only create one room per one worker.

**Reply by gegaojian-richard** [reply]
thank you very much!
